File: packages/harmony-audit/src/runtime/evidence.ts
/** Evidence ownership and admissibility for AI task boundaries (v3.1-aligned inline model). */
import { canonicalJson, contentHash } from "./identity.js";

export type Row = Record<string, unknown>;
const rows = (value: unknown): Row[] => Array.isArray(value) ? value.filter((item): item is Row => !!item && typeof item === "object" && !Array.isArray(item)) : [];
const strings = (value: unknown): string[] => Array.isArray(value) ? value.filter((item): item is string => typeof item === "string") : [];
const unique = (values: readonly string[]): string[] => [...new Set(values)].sort();
const isRow = (value: unknown): value is Row => !!value && typeof value === "object" && !Array.isArray(value);

export type EvidenceRow = Row & { kind: string; source: string; summary: string };

const EVIDENCE_FIELDS = ["kind", "source", "summary", "location", "content_ref", "sha256"] as const;

export function evidenceRow(value: unknown): EvidenceRow {
  return Object.fromEntries(EVIDENCE_FIELDS
    .filter((key) => isRow(value) && value[key] !== undefined)
    .map((key) => [key, isRow(value) ? value[key] : undefined])) as EvidenceRow;
}

/** Content-addressed local id: identical evidence rows collapse to one id per producer task. */
export function evidenceLocalId(row: EvidenceRow): string { return contentHash(canonicalJson(row)); }

export type EvidenceCollector = Map<string, EvidenceRow>;

function register(out: EvidenceCollector, items: unknown): string[] {
  const ids: string[] = [];
  for (const item of rows(items)) {
    const row = evidenceRow(item);
    const id = evidenceLocalId(row);
    out.set(id, row);
    ids.push(id);
  }
  return unique(ids);
}

/** Recursively replace inline `evidence` arrays with runtime-numbered `evidence_refs`. */
function materializeInline(value: unknown, out: EvidenceCollector): unknown {
  if (Array.isArray(value)) return value.map((item) => materializeInline(item, out));
  if (!isRow(value)) return value;
  const result: Row = {};
  for (const [key, item] of Object.entries(value)) {
    if (key === "evidence" && Array.isArray(item)) result.evidence_refs = register(out, item);
    else if (key === "basis_evidence" && Array.isArray(item)) result.basis_evidence_refs = register(out, item);
    else result[key] = materializeInline(item, out);
  }
  return result;
}

/** Source-backed evidence that may support a security conclusion. */
export function semanticAdmissibleRefs(group: Row): Set<string> {
  const refs = [
    ...strings(group.evidence_refs),
    ...strings(isRow(group.operation) ? group.operation.evidence_refs : []),
  ];
  for (const key of ["facts", "edges", "branches", "security_checks"]) {
    for (const item of rows(group[key])) refs.push(...strings(item.evidence_refs));
  }
  const context = isRow(group.context) ? group.context : {};
  const availability = isRow(group.availability) ? group.availability : {};
  refs.push(...strings(context.evidence_refs), ...strings(availability.evidence_refs));
  return new Set(refs);
}

/** Hypothesis basis evidence that is not admissible: untrusted for validation conclusions. */
export function semanticHypothesisRefs(group: Row): Set<string> {
  const context = isRow(group.context) ? group.context : {};
  const refs: string[] = [];
  for (const hypothesis of rows(context.effect_hypotheses)) refs.push(...strings(hypothesis.basis_evidence_refs));
  const admissible = semanticAdmissibleRefs(group);
  return new Set(refs.filter((ref) => !admissible.has(ref)));
}

export function componentCallRefs(call: Row): Set<string> {
  const refs = [...strings(call.evidence_refs)];
  const transition = isRow(call.principal_transition) ? call.principal_transition : {};
  refs.push(...strings(transition.evidence_refs));
  for (const check of rows(call.security_checks)) refs.push(...strings(check.evidence_refs));
  return new Set(refs);
}

export function materializeSemanticGroup(group: Row, out: EvidenceCollector): Row {
  const materialized = materializeInline(group, out) as Row;
  const facts = rows(materialized.facts);
  materialized.edges = facts.slice(0, -1).map((fact, index) => ({
    from: String(fact.fact_key),
    to: String(facts[index + 1]!.fact_key),
    kind: "next",
    evidence_refs: unique([...strings(fact.evidence_refs), ...strings(facts[index + 1]!.evidence_refs)]),
  }));
  materialized.evidence_refs = [...semanticAdmissibleRefs(materialized)].sort();
  return materialized;
}

export function materializeComponentCall(call: Row, out: EvidenceCollector): Row {
  const materialized = materializeInline(call, out) as Row;
  materialized.evidence_refs = [...componentCallRefs(materialized)].sort();
  return materialized;
}

export function materializeValidation(validation: Row, out: EvidenceCollector): Row {
  function convert(value: unknown): unknown {
    if (Array.isArray(value)) return value.map(convert);
    if (!isRow(value)) return value;
    const result: Row = {};
    for (const [key, item] of Object.entries(value)) {
      if (key === "evidence" && isRow(item) && "semantic_refs" in item && "verification" in item) {
        const semantic = strings(item.semantic_refs);
        const verification = register(out, item.verification);
        result.evidence_refs = unique([...semantic, ...verification]);
      } else {
        result[key] = convert(item);
      }
    }
    return result;
  }
  return convert(validation) as Row;
}

/** All semantic refs a validation cites across every `evidence: {semantic_refs, verification}` support. */
export function validationSemanticRefs(value: unknown): Set<string> {
  const refs = new Set<string>();
  const walk = (item: unknown): void => {
    if (Array.isArray(item)) { item.forEach(walk); return; }
    if (!isRow(item)) return;
    const support = item.evidence;
    if (isRow(support) && "semantic_refs" in support && "verification" in support) {
      strings(support.semantic_refs).forEach((ref) => refs.add(ref));
    }
    for (const [key, child] of Object.entries(item)) if (key !== "evidence") walk(child);
  };
  walk(value);
  return refs;
}

export function materializePoc(poc: Row, out: EvidenceCollector): Row {
  const materialized = structuredClone(poc) as Row;
  const materializedRefs: string[] = [];
  for (const symbolRef of rows(materialized.symbol_refs)) {
    const refs = register(out, symbolRef.evidence);
    if (refs.length) {
      symbolRef.evidence_refs = refs;
      delete symbolRef.evidence;
      materializedRefs.push(...refs);
    }
  }
  materialized.evidence_refs = unique([...strings(materialized.evidence_refs), ...materializedRefs]);
  return materialized;
}

File: resources/skills/audit-orchestration/scripts/audit_runtime/evidence.py
"""Evidence ownership and admissibility for AI task boundaries."""
from __future__ import annotations

import json

from .common import canonical_json, now, stable_id


EVIDENCE_FIELDS = ("kind", "source", "summary", "location", "content_ref", "sha256")


def _evidence_row(row):
    return {key: row.get(key) for key in EVIDENCE_FIELDS if key in row}


def _insert_inline_evidence(conn, task_id, rows, role):
    evidence_ids = []
    for source in rows or []:
        row = _evidence_row(source)
        evidence_id = stable_id("EVID", [task_id, role, canonical_json(row)])
        conn.execute(
            """INSERT OR IGNORE INTO evidence
               (evidence_id,task_id,kind,source,location,summary,content_ref,sha256,created_at)
               VALUES (?,?,?,?,?,?,?,?,?)""",
            (evidence_id, task_id, row["kind"], row["source"], row.get("location"), row["summary"],
             row.get("content_ref"), row.get("sha256"), now()),
        )
        evidence_ids.append(evidence_id)
    return sorted(set(evidence_ids))


def _materialize_inline_evidence(conn, task_id, value):
    if isinstance(value, list):
        return [_materialize_inline_evidence(conn, task_id, item) for item in value]
    if not isinstance(value, dict):
        return value
    result = {}
    for key, item in value.items():
        if key == "evidence" and isinstance(item, list):
            result["evidence_refs"] = _insert_inline_evidence(conn, task_id, item, "semantic")
        elif key == "basis_evidence" and isinstance(item, list):
            result["basis_evidence_refs"] = _insert_inline_evidence(conn, task_id, item, "hypothesis")
        else:
            result[key] = _materialize_inline_evidence(conn, task_id, item)
    return result


def semantic_admissible_refs(group):
    """Return source-backed evidence that may support a security conclusion."""
    refs = list(group.get("evidence_refs", []))
    refs.extend(group.get("operation", {}).get("evidence_refs", []))
    for key in ("facts", "edges", "branches", "security_checks"):
        for row in group.get(key, []):
            refs.extend(row.get("evidence_refs", []))
    refs.extend(group.get("context", {}).get("evidence_refs", []))
    refs.extend(group.get("availability", {}).get("evidence_refs", []))
    return set(refs)


def semantic_hypothesis_refs(group):
    refs = []
    for hypothesis in group.get("context", {}).get("effect_hypotheses", []):
        refs.extend(hypothesis.get("basis_evidence_refs", []))
    return set(refs) - semantic_admissible_refs(group)


def component_call_refs(component_call):
    refs = list(component_call.get("evidence_refs", []))
    refs.extend(component_call.get("invocation_control", {}).get("evidence_refs", []))
    refs.extend(component_call.get("principal_transition", {}).get("evidence_refs", []))
    for security_check in component_call.get("security_checks", []):
        refs.extend(security_check.get("evidence_refs", []))
    return set(refs)


def materialize_semantic_group(conn, task_id, source):
    group = _materialize_inline_evidence(conn, task_id, json.loads(json.dumps(source)))
    facts = group.get("facts", [])
    group["edges"] = [{
        "from": left["fact_key"],
        "to": right["fact_key"],
        "kind": "next",
        "evidence_refs": sorted(set(left.get("evidence_refs", [])) | set(right.get("evidence_refs", []))),
    } for left, right in zip(facts, facts[1:])]
    group["evidence_refs"] = sorted(semantic_admissible_refs(group))
    return group


def materialize_component_call(conn, task_id, source):
    component_call = _materialize_inline_evidence(conn, task_id, json.loads(json.dumps(source)))
    component_call["evidence_refs"] = sorted(component_call_refs(component_call))
    return component_call


def iter_validation_supports(value):
    if isinstance(value, list):
        for item in value:
            yield from iter_validation_supports(item)
        return
    if not isinstance(value, dict):
        return
    support = value.get("evidence")
    if isinstance(support, dict) and "semantic_refs" in support and "verification" in support:
        yield support
    for key, item in value.items():
        if key != "evidence":
            yield from iter_validation_supports(item)


def validation_semantic_refs(value):
    return {
        ref
        for support in iter_validation_supports(value)
        for ref in support.get("semantic_refs", [])
    }


def materialize_validation(conn, task_id, source):
    def convert(value):
        if isinstance(value, list):
            return [convert(item) for item in value]
        if not isinstance(value, dict):
            return value
        result = {}
        for key, item in value.items():
            if key == "evidence" and isinstance(item, dict):
                semantic_refs = item.get("semantic_refs", [])
                verification_refs = _insert_inline_evidence(
                    conn, task_id, item.get("verification", []), "verification"
                )
                result["evidence_refs"] = sorted(set(semantic_refs) | set(verification_refs))
            else:
                result[key] = convert(item)
        return result

    return convert(json.loads(json.dumps(source)))


def materialize_poc(conn, task_id, source):
    """PoC artifacts use the same inline-evidence model: each symbol_ref carries
    its own evidence without IDs; the runtime numbers, dedupes, and merges refs.
    Top-level evidence_refs may only reference inherited evidence ids."""
    poc = json.loads(json.dumps(source))
    materialized_refs = []
    for symbol_ref in poc.get("symbol_refs", []):
        if isinstance(symbol_ref, dict) and isinstance(symbol_ref.get("evidence"), list):
            refs = _insert_inline_evidence(conn, task_id, symbol_ref["evidence"], "poc")
            symbol_ref["evidence_refs"] = sorted(refs)
            symbol_ref.pop("evidence", None)
            materialized_refs.extend(refs)
    poc["evidence_refs"] = sorted(set(poc.get("evidence_refs", [])) | set(materialized_refs))
    poc["assurance_status"] = "generated_unverified"
    return poc
